fuse/man/fuse.1
.\" -*- nroff -*-
.\"
.\" fuse.1: Fuse man page
.\" Copyright (c) 2001-2018 Russell Marks, Philip Kendall, Darren Salt,
.\"                         Fredrick Meunier, Stuart Brady
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Author contact information:
.\"
.\" E-mail: philip-fuse@shadowmagic.org.uk
.\"
.\"
.TH fuse 1 "9th December, 2018" "Version 1.5.7" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
fuse \(em Sinclair ZX Spectrum emulator
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.\" the trend for long-option-using programs is to give a largely
.\" generic synopsis, so...
.B fuse
.RI [ options ]
.P
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
Fuse is a Sinclair ZX\ Spectrum emulator. It supports several models
(including the 128), with quite faithful emulation of the display and
sound.
.PP
The emulator can load any of the formats supported by
.IR libspectrum (3)
\(em this includes Z80, SNA and SZX snapshots, and PZX, TAP and TZX
virtual-tape files. Saving to SZX, Z80 and SNA snapshots and TZX and
TAP tape files is supported. The SLT extension to the Z80 format is
partly supported (enough for multi-load games); however, loading of
the old DAT-file variant is not.
.PP
DSK, UDI, FDI, TD0, MGT, IMG, D40, D80, SAD, TRD, SCL and OPD
disk images are supported when a disk interface is being emulated,
including the integrated disk drives on +3, Pentagon or Scorpion
machines as well as the +D, DISCiPLE, Opus Discovery, Beta\ 128 and
Didaktik\ 80 interfaces. DCK cartridge images are supported when
emulating a Timex\ 2068 variant. Interface\ 2 ROM cartridges are
also supported.
.PP
Finally, there is also support for reading and writing the RZX input recording
format.
.PP
See the
.B "COMPRESSED FILES"
section for details on reading files compressed with
.IR bzip2 "(3),"
.IR gzip "(3)"
or
.IR zip "(3)."
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.\" dual short/long option listings here reflect the GNU approach,
.\" as used in info files. It does mean using RS/RE, though,
.\" so this is used for all options (for consistency).
.\"
.\" the options list is in alphabetical order by long option name (or
.\" short option name if none).
.\"
.B \-\-accelerate\-loader
.RS
Specify whether Fuse should attempt to accelerate tape loaders by \(lqshort
circuiting\(rq the loading loop. This will in general speed up loading, but
may cause some loaders to fail. (Enabled by default, but you can use
.RB ` \-\-no\-accelerate\-loader '
to disable). The same as the Media Options dialog's
.I "Accelerate loaders"
option.
.RE
.PP
.B \-\-aspect\-hint
.RS
Specify whether the GTK+ and Xlib user interfaces should `hint' to the
window manager about the preferred aspect ratio for the graphics
window, thus preventing resizing to non-square sizes which lead to
Fuse not displaying correctly. This option has been observed to cause
problems with some window managers when using the GTK+ UI which can
prevent the window from being resized or moved at all. (Enabled by
default, but you can use
.RB ` \-\-no\-aspect\-hint '
to disable). See also the
.RB ` \-\-strict\-aspect\-hint '
option.
.RE
.PP
.B \-\-autosave\-settings
.RS
Specify whether Fuse's current settings should be automatically saved
on exit. The same as the General Options dialog's
.I "Auto-save settings"
option.
.RE
.PP
.B \-\-auto\-load
.RS
Specify whether tape and disk files should be automatically loaded
when they are opened using the
.I "File, Open..."
menu option. In the case of TRD/SCL disk images, inserts also a boot
loader file when none is available. (Enabled by default, but you can use
.RB ` \-\-no\-auto\-load '
to disable). Same as the Media Options dialog's
.I "Auto-load media"
option.
.RE
.PP
.B \-\-beta128
.RS
Emulate a Beta\ 128 interface. Same as the Disk Peripherals Options dialog's
.I "Beta\ 128 interface"
option.
.RE
.PP
.B \-\-beta128\-48boot
.RS
When a Beta\ 128 interface is used in 48K or TC2048 emulation the
option additionally controls whether the machine boots directly into the TR-DOS
system. Same as the Disk Peripherals Options dialog's
.I "Beta\ 128 auto-boot in 48K machines"
option.
.RE
.PP
.B \-\-betadisk
.I file
.RS
Insert the specified file into the emulated Beta disk interface's drive\ A:
and select Pentagon mode on startup.
.RE
.PP
.B \-\-bw\-tv
.RS
Specify whether the display should simulate a colour or black and
white television. This option is effective under the GTK+, Win32, Xlib
and SDL user interfaces: the others will always simulate a colour TV.
The same as the General Options dialog's
.I "Black and white TV"
option.
.RE
.PP
.B \-\-cmos-z80
.RS
This option specifies that Fuse should emulate a CMOS Z80, as opposed
to an NMOS Z80. Same as the General Options dialog's
.I "Z80 is CMOS"
option.
.RE
.PP
.B \-\-competition\-code
.I code
.RS
Specify the code to be written to competition mode RZX files. The same
as the RZX Options dialog's
.I "Competition code"
option.
.RE
.PP
.B \-\-competition\-mode
.RS
Specify whether input recordings should be made in `competition mode'.
The same as the RZX Options dialog's
.I "Competition mode"
option.
.RE
.PP
.B \-\-compress\-rzx
.RS
Specify whether RZX files should be written out compressed. (Enabled
by default, but you can use
.RB ` \-\-no\-compress\-rzx '
to disable). Same as the RZX Options dialog's
.I "Compress RZX data"
option.
.RE
.PP
.B \-\-confirm\-actions
.RS
Specify whether `dangerous' actions (those which could cause data
loss, for example resetting the Spectrum) require confirmation before
occurring. (Enabled by default, but you can use
.RB ` \-\-no\-confirm\-actions '
to disable). This option is the same as the General Options dialog's
.I "Confirm actions"
option.
.RE
.PP
.B \-\-covox
.RS
Emulate a Covox sound interface for Pentagon/Scorpion. Same as the
General Peripherals Options dialog's
.I "Covox"
option.
.RE
.PP
.B \-\-debugger\-command
.I string
.RS
Specify a debugger command to be run before emulator startup. This can
be used to set breakpoints or the like. Currently, this is the only
method to input multi-line debugger commands. (See the
.B "MONITOR/DEBUGGER"
section for more information).
.RE
.PP
.B \-\-detect\-loader
.RS
Specify whether Fuse should attempt to detect when the tape is being
accessed and start and stop the virtual tape playing
automatically. (Enabled by default, but you can use
.RB ` \-\-no\-detect\-loader '
to disable). Same as the Media Options dialog's
.I "Detect loaders"
option.
.RE
.PP
.B \-\-disciple
.RS
Emulate a DISCiPLE interface. Same as the Disk Peripherals Options dialog's
.I "DISCiPLE interface"
option.
.RE
.PP
.B \-\-discipledisk
.I file
.RS
Insert the specified file into the emulated DISCiPLE's drive\ 1.
.RE
.PP
.B \-\-didaktik80
.RS
Emulate a Didaktik 80 (or Didaktik 40) disk interface. Same as the
Disk Peripherals Options dialog's
.I "Didaktik 80 interface"
option.
.RE
.PP
.B \-\-didaktik80disk
.I file
.RS
Insert the specified file into the emulated Didaktik\ 80
(or Didaktik\ 40)'s drive\ A.
.RE
.PP
.B \-\-disk\-ask\-merge
.RS
Prompt the user to confirm whether Fuse should try to merge the `B' side of a
disk image from a separate file when opening a new single-sided disk image.
.RE
.PP
.B \-\-disk\-try\-merge
.I mode
.RS
Select whether Fuse should try to merge a separate file for the `B' side of a
disk image separate file when opening a new disk image. Most double sided disk
images are dumped as two single sided disk images e.g.
`Golden Axe \- Side\ A.dsk' and `Golden Axe \- Side\ B.dsk'. So, if we want to
play Golden Axe, first we have to insert the first disk image and when the game
asks to insert side\ B, we have to find and open the second disk image, instead
of just `flip'-ing the disk inside the drive. If enabled, Fuse will try to open
the second image too and create a double sided disk image (merging the two one
sided disk images) and insert this merged virtual disk into the disk drive. The
function detects whether the file is one side of a double-sided image if the
filename matches a pattern like [Ss]ide[ _][abAB12][ _.] in the file name of a
disk that is being opened. If found, Fuse will try to open the other side of
the disk too substituting the appropriate characters in the filename e.g.
1\(->2, a\(->b, A\(->B. If successful then it will merge the two images and
now we have a double sided disk in drive. This means that if we open
`Golden Axe \- Side\ A.dsk', then Fuse will try to open `Golden Axe \- Side
B.dsk' too. Now, we can just `flip' the disk if Golden Axe asks for `Side\ B'.
The available options are
.IR Never ,
.I "With single-sided drives"
and
.IR Always .
.RE
.PP
.B \-\-divide
.RS
Emulate the DivIDE interface. The same as the Disk Peripherals Options
dialog's
.I "DivIDE interface"
option.
.RE
.PP
.B \-\-divide\-masterfile
.I file
.br
.B \-\-divide\-slavefile
.I file
.RS
Specify an IDE image to be loaded into the DivIDE's emulated master
and slave drives respectively.
.RE
.PP
.B \-\-divide\-write\-protect
.RS
Specify that the emulated DivIDE's write protect jumper should be
considered set. The same as the Disk Peripherals Options dialog's
.I "DivIDE write protect"
option.
.RE
.PP
.B \-\-divmmc
.RS
Emulate the DivMMC interface. The same as the Disk Peripherals Options
dialog's
.I "DivMMC interface"
option.
.RE
.PP
.B \-\-divmmc\-file
.I file
.RS
Specify an HDF image to be loaded into the DivMMC's emulated memory card.
.RE
.PP
.B \-\-divmmc\-write\-protect
.RS
Specify that the emulated DivMMC's write protect jumper that protects
EEPROM should be considered set. The same as the Disk Peripherals Options
dialog's
.I "DivMMC write protect"
option.
.RE
.PP
.B \-\-dock
.I file
.RS
Insert the specified file into the emulated Timex\ 2068 variant dock;
also select the TC2068 on startup if available.
.RE
.PP
.B \-D
.I mode
.br
.B \-\-doublescan\-mode
.I mode
.RS
Specify whether to use doublescan modes in the FB UI.
Available values for
.I mode
are 0, 1 and 2. 0 means `never doublescan' (use 640\(mu480 at either 72\ Hz or
60\ Hz), whereas 1 and 2 both mean `try to use doublescan' and will fall back
on the 640\(mu480 modes. 1 selects 72\ Hz modes (the same size and shape as
your typical 640\(mu480), and 2 selects 60\ Hz modes (overscan).
.PP
If your monitor displays a blank screen when using 1 or 2, press F10 then
try a different option or say `\-\-fbmode 640'.
.RE
.PP
.B \-\-drive\-plus3a\-type
.I type
.br
.B \-\-drive\-plus3b\-type
.I type
.br
.B \-\-drive\-beta128a\-type
.I type
.br
.B \-\-drive\-beta128b\-type
.I type
.br
.B \-\-drive\-beta128c\-type
.I type
.br
.B \-\-drive\-beta128d\-type
.I type
.br
.B \-\-drive\-plusd1\-type
.I type
.br
.B \-\-drive\-plusd2\-type
.I type
.br
.B \-\-drive\-didaktik80a\-type
.I type
.br
.B \-\-drive\-didaktik80b\-type
.I type
.br
.B \-\-drive\-disciple1\-type
.I type
.br
.B \-\-drive\-disciple2\-type
.I type
.br
.B \-\-drive\-opus1\-type
.I type
.br
.B \-\-drive\-opus2\-type
.I type
.RS
Specify a disk drive type to emulate with the associated interface. 
The available options are
.IR "Disabled" ,
.IR "Single\-sided 40 track" ,
.IR "Double\-sided 40 track" ,
.IR "Single\-sided 80 track" " and"
.IR "Double\-sided 80 track" .
See the Disk Options dialog for more information.  The
.I Disabled
option is not supported for Drive 1 or Drive A of any interface.
.RE
.PP
.B \-\-drive\-40\-max\-track
.I count
.br
.B \-\-drive\-80\-max\-track
.I count
.RS
Specify the maximum number of tracks for 40 and 80 track physical drives
respectively.
.RE
.PP
.B \-\-embed\-snapshot
.RS
Specify whether a snapshot should be embedded in an RZX file when
recording is started from an existing snapshot. (Enabled by default,
but you can use
.RB ` \-\-no\-embed\-snapshot '
to disable). Same as the RZX Options dialog's
.I "Always embed snapshot"
option.
.RE
.PP
.B \-\-fastload
.RS
Specify whether Fuse should run at the fastest possible speed when the
virtual tape is playing. (Enabled by default, but you can use
.RB ` \-\-no\-fastload '
to disable). The same as the Media Options dialog's
.I "Fastloading"
option.
.RE
.PP
.B \-v
.I mode
.br
.B \-\-fbmode
.I mode
.RS
Specify which mode to use for the FB UI. Available values for
.I mode
are `320' (which corresponds to a 320\(mu240\(mu256 mode), the default and
`640' (a 640\(mu480\(mu256 mode).
.RE
.PP
.B \-\-fuller
.RS
Emulate a Fuller Box interface. Same as the General Peripherals Options dialog's
.I "Fuller Box"
option.
.RE
.PP
.B \-\-full\-screen
.RS
Specify whether Fuse should run in full screen mode.
This option is effective only under the SDL UI.
.RE
.PP
.B \-g
.I filter
.br
.B \-\-graphics\-filter
.I mode
.RS
Specify which graphics filter to use if available. The default is
.IR normal ,
which uses no filtering. The available options are
.IR 2x ,
.IR 2xsai ,
.IR 3x ,
.IR advmame2x ,
.IR advmame3x ,
.IR dotmatrix ,
.IR half ,
.IR halfskip ,
.IR hq2x ,
.IR hq3x ,
.IR normal ,
.IR super2xsai ,
.IR supereagle ,
.IR timex15x ,
.IR timextv ,
.IR tv2x ,
.IR paltv ,
.IR paltv2x ,
and
.IR paltv3x .
See the
.B "GRAPHICS FILTERS"
section for more details.
.RE
.PP
.B \-\-graphicsfile
.I file
.RS
Set the filename used for graphical output from the emulated
ZX\ Printer. See the
.B "PRINTER EMULATION"
section for more details.
.RE
.PP
.B \-h
.br
.B \-\-help
.RS
Give brief usage help, listing available options.
.RE
.PP
.B \-\-if2cart
.I file
.RS
Insert the specified file into the emulated Interface\ 2.
.RE
.PP
.B \-\-interface1
.RS
Emulate a Sinclair Interface\ 1. Same as the General Peripherals Options dialog's
.I "Interface\ 1"
option.
.RE
.PP
.B \-\-interface2
.RS
Emulate a Sinclair Interface\ 2. (Enabled by default, but you can use
.RB ` \-\-no\-interface2 '
to disable). Same as the General Peripherals Options dialog's
.I "Interface\ 2"
option.
.RE
.PP
.B \-\-issue2
.RS
Emulate an issue\ 2 keyboard. Same as the General Options dialog's
.I "Issue\ 2 keyboard"
option.
.RE
.PP
.B \-j
.I device
.br
.B \-\-joystick\-1
.I device
.RS
Read from
.I device
to emulate the first joystick. Fuse will use either
.RI ` /dev/input/js0 '
or
.RI ` /dev/js0 '
by default.
.RE
.PP
.B \-\-joystick\-2
.I device
.RS
As for
.B \-\-joystick\-1
but for the second joystick; the default here is either
.RI ` /dev/input/js1 '
or
.RI ` /dev/js1 "'."
.RE
.PP
.B \-\-joystick\-1\-output
.I type
.br
.B \-\-joystick\-2\-output
.I type
.br
.B \-\-joystick\-keyboard\-output
.I type
.RS
Select which joystick interface to attach for the first two real
joysticks and the keyboard joystick. The default is
.IR 0 ,
which is no output. The available options are
.I 1
(cursor),
.I 2
(kempston),
.I 3
(Sinclair\ 1),
.I 4
(Sinclair\ 2),
.I 5
(Timex\ 1),
.I 6
(Timex\ 2),
and
.I 7
(Fuller). Same as the Joysticks Options dialog's
.I "Joystick type"
option.
.RE
.PP
.B \-\-joystick\-1\-fire\-1
.I code
.br
.B \-\-joystick\-1\-fire\-2
.I code
.br
.B \-\-joystick\-1\-fire\-3
.I code
.br
.B \-\-joystick\-1\-fire\-4
.I code
.br
.B \-\-joystick\-1\-fire\-5
.I code
.br
.B \-\-joystick\-1\-fire\-6
.I code
.br
.B \-\-joystick\-1\-fire\-7
.I code
.br
.B \-\-joystick\-1\-fire\-8
.I code
.br
.B \-\-joystick\-1\-fire\-9
.I code
.br
.B \-\-joystick\-1\-fire\-10
.I code
.br
.B \-\-joystick\-1\-fire\-11
.I code
.br
.B \-\-joystick\-1\-fire\-12
.I code
.br
.B \-\-joystick\-1\-fire\-13
.I code
.br
.B \-\-joystick\-1\-fire\-14
.I code
.br
.B \-\-joystick\-1\-fire\-15
.I code
.br
.B \-\-joystick\-2\-fire\-1
.I code
.br
.B \-\-joystick\-2\-fire\-2
.I code
.br
.B \-\-joystick\-2\-fire\-3
.I code
.br
.B \-\-joystick\-2\-fire\-4
.I code
.br
.B \-\-joystick\-2\-fire\-5
.I code
.br
.B \-\-joystick\-2\-fire\-6
.I code
.br
.B \-\-joystick\-2\-fire\-7
.I code
.br
.B \-\-joystick\-2\-fire\-8
.I code
.br
.B \-\-joystick\-2\-fire\-9
.I code
.br
.B \-\-joystick\-2\-fire\-10
.I code
.br
.B \-\-joystick\-2\-fire\-11
.I code
.br
.B \-\-joystick\-2\-fire\-12
.I code
.br
.B \-\-joystick\-2\-fire\-13
.I code
.br
.B \-\-joystick\-2\-fire\-14
.I code
.br
.B \-\-joystick\-2\-fire\-15
.I code
.RS
Select which Fuse key code should be triggered by the applicable
real joystick button press. The codes are the Fuse keyboard codes
corresponding to the keys. The default value is
.I 4096
which corresponds to the virtual joystick fire button. Same as the
Joysticks Options dialog's
.I "Joystick fire"
options.
.RE
.PP
.B \-\-joystick\-keyboard\-up
.I code
.br
.B \-\-joystick\-keyboard\-down
.I code
.br
.B \-\-joystick\-keyboard\-left
.I code
.br
.B \-\-joystick\-keyboard\-right
.I code
.br
.B \-\-joystick\-keyboard\-fire
.I code
.RS
Select which Fuse key code should correspond with each direction and
fire for the keyboard virtual joystick. The same as the Keyboard
Joysticks Options dialog's
.IR "Button for UP" ,
.IR "Button for DOWN" ,
.IR "Button for LEFT" ,
.I "Button for RIGHT"
and
.I "Button for FIRE"
options respectively.
.RE
.PP
.B \-\-joystick\-prompt
.RS
If this option is specified, then Fuse will prompt you which form of
joystick emulation you wish to use when loading a snapshot. No prompt
will be issued if the configuration in the snapshot matches what you
are currently using. The same as the General Options dialog's
.I "Snap joystick prompt"
option.
.RE
.PP
.B \-\-kempston
.RS
Emulate a Kempston joystick. Same as the General Peripherals Options dialog's
.I "Kempston joystick"
option.
.RE
.PP
.B \-\-kempston\-mouse
.RS
Emulate a Kempston mouse. Same as the General Peripherals Options dialog's
.I "Kempston mouse"
option.
.RE
.PP
.B \-\-keyboard\-arrows\-shifted
.RS
Treat the keyboard arrow keys as shifted like the ZX Spectrum+ keyboard's
arrow keys or as unshifted like a cursor joystick that maps to the 5, 6, 7
and 8 keys. (Enabled by default, but you can use
.RB ` \-\-no\-keyboard\-arrows\-shifted '
to disable). Same as the General Peripherals Options dialog's
.I "Use shift with arrow keys"
option.
.RE
.PP
.B \-\-late\-timings
.RS
It has been observed that some real Spectrums run such that the screen
is rendered one tstate later than on other real hardware. This option
specifies that Fuse should emulate such a machine. Same as the General
Options dialog's
.I "Late timings"
option.
.RE
.PP
.B \-\-loading\-sound
.RS
Specify whether the sound made while tapes are loading should be
emulated. (Enabled by default, but you can use
.RB ` \-\-no\-loading\-sound '
to disable). Same as the Sound Options dialog's
.I "Loading sound"
option.
.RE
.PP
.B \-m
.I type
.br
.B \-\-machine
.I type
.RS
Specify machine type to emulate initially. The default is
.IR 48 ,
a 48K\ Spectrum. The available options are
.IR 16 ,
.IR 48 ,
.IR 48_ntsc ,
.IR 128 ,
.IR plus2 ,
.IR plus2a ,
.IR plus3 ,
.IR 2048 ,
.IR 2068 ,
.IR ts2068 ,
.IR pentagon ,
.IR pentagon512 ,
.IR pentagon1024 ,
.I scorpion
and
.IR se .
.RE
.PP
.B \-\-melodik
.RS
Emulate a Melodik AY\ interface for 16/48k\ Spectrums. Same as the
General Peripherals Options dialog's
.I "Melodik"
option.
.RE
.PP
.B \-\-mdr\-len
.I length
.RS
This option controls the number of blocks in a new Microdrive cartridge.
Same as the Media Options dialog's
.I "MDR cartridge len"
option.
.RE
.PP
.B \-\-mdr\-random\-len
.RS
If this option is set, Fuse will use a random Microdrive cartridge
length. Same as the Media Options dialog's
.I "Random length MDR cartridge"
option.
.RE
.PP
.B \-\-microdrive\-file
.I file
.br
.B \-\-microdrive\-2\-file
.I file
.br
.B \-\-microdrive\-3\-file
.I file
.br
.B \-\-microdrive\-4\-file
.I file
.br
.B \-\-microdrive\-5\-file
.I file
.br
.B \-\-microdrive\-6\-file
.I file
.br
.B \-\-microdrive\-7\-file
.I file
.br
.B \-\-microdrive\-8\-file
.I file
.RS
Specify Interface\ 1 Microdrive cartridge files to open.
.RE
.PP
.B \-\-mouse\-swap\-buttons
.RS
Swap the left and right mouse buttons when emulating the Kempston mouse. The
same as the General Peripherals dialog's
.I "Swap mouse buttons"
option.
.RE
.PP
.B \-\-movie\-compr
.I level
.RS
This option sets the compression level used when creating movies. Same as the
Movie Options dialog's
.I "Movie compression"
option. The available options are
.IR None ,
.IR Lossless ,
and
.I High
(lossy). The default option is
.IR Lossless .
See also the
.B "MOVIE RECORDING"
section.
.RE
.PP
.B \-\-movie\-start
.I filename
.RS
With this command line option, Fuse will start movie recording as soon as the
emulator is started. See also the
.B "MOVIE RECORDING"
section.
.RE
.PP
.B \-\-movie\-stop\-after\-rzx
.RS
With this command line option, Fuse will stop movie recording when RZX playback
or RZX recording ends. Same as the Movie Options dialog's
.I "Stop recording after RZX ends"
option. (Enabled by default, but you can use
.RB ` \-\-no\-movie\-stop\-after\-rzx '
to disable).
See also the
.B "MOVIE RECORDING"
section.
.RE
.PP
.B \-\-multiface1
.RS
Emulate a Romantic Robot Multiface One interface. Same as the General
Peripherals Options dialog's
.I "Multiface One"
option.
.RE
.PP
.B \-\-multiface128
.RS
Emulate a Romantic Robot Multiface\ 128 interface. Same as the General
Peripherals Options dialog's
.I "Multiface\ 128"
option.
.RE
.PP
.B \-\-multiface3
.RS
Emulate a Romantic Robot Multiface\ 3 interface. Same as the General
Peripherals Options dialog's
.I "Multiface\ 3"
option.
.RE
.PP
.B \-\-multiface1-stealth
.RS
Set Multiface One stealth/invisible mode. Same as the General Peripherals
Options dialog's
.I "Stealth Multiface One"
option.
.RE
.PP
.B \-\-opus
.RS
Emulate an Opus Discovery interface. Same as the Disk Peripherals Options
dialog's
.I "Opus Discovery interface"
option.
.RE
.PP
.B \-\-opusdisk
.I file
.RS
Insert the specified file into the emulated Opus Discovery's drive\ 1.
.RE
.PP
.B \-\-pal\-tv2x
.RS
Specify whether the PAL\ TV\ 2x and PAL\ TV\ 3x scalers should also produce
scanlines along the lines of the TV\ 2x and Timex\ TV scalers.
The same as the General Options dialog's
.I "PAL-TV use TV2x effect"
option.
.RE
.PP
.B \-\-phantom\-typist\-mode
.I mode
.RS
Specify the keystroke sequence that the "phantom typist" should use when
starting a program loading. The available options are
.IR "Auto" ,
.IR "Keyword" ,
.IR "Keystroke" ,
.IR "Menu" ,
.IR "Plus 2A" " and"
.IR "Plus 3" .
The same as the Media Options dialog's
.I "Phantom typist mode"
option.
.RE
.PP
.B \-p
.I file
.br
.B \-\-playback
.I file
.RS
Specify an RZX file to begin playback from.
.RE
.PP
.B \-\-plus3disk
.I file
.RS
Insert the specified file into the emulated +3's A:\ drive; also select
the +3 on startup if available.
.RE
.PP
.B \-\-plus3\-detect\-speedlock
.RS
Specify whether the +3 drives try to detect Speedlock protected disks,
and emulate `weak' sectors.
If the disk image file (EDSK or UDI) contains weak sector data, than
Speedlock detection is automatically omitted.
See also the
.B "WEAK DISK DATA"
section.
Same as the Disk Options dialog's
.I "+3 Detect Speedlock"
option.
.RE
.PP
.B \-\-plusd
.RS

FuseZX/Fuse.swift
/*
 FuseZX.swift -- High Level Interface to fuse
 Copyright (C) 2019 Dieter Baron
 
 This file is part of C64, a Commodore 64 emulator for iOS, based on VICE.
 The authors can be contacted at <c64@spiderlab.at>
 
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 02111-1307  USA.
 */

import UIKit
import Emulator

import FuseC

@objc public class Fuse: Emulator {
    public override init() {
        fuseThread = FuseThread()
        super.init()
        fuseThread?.delegate = self
    }
    
    private var pressedKeys = Multiset<Key>()
    
    func fuseName(for model: Computer.ViceModel) -> String? {
        switch model {
        case .zxSpectrum16k:
            return "16"
            
        case .zxSpectrum48k:
            return "48"
            
        case .zxSpectrum48kNtsc:
            return "48_ntsc"
            
        case .zxSpectrum128k:
            return "128"
            
        default:
            return nil
        }
    }
    
    func fuseKeys(for key: Key) -> [keyboard_key_name]? {
        switch key {
        case .Char(" "):
            return [KEYBOARD_space]
        case .Char("0"):
            return [KEYBOARD_0]
        case .Char("1"):
            return [KEYBOARD_1]
        case .Char("2"):
            return [KEYBOARD_2]
        case .Char("3"):
            return [KEYBOARD_3]
        case .Char("4"):
            return [KEYBOARD_4]
        case .Char("5"):
            return [KEYBOARD_5]
        case .Char("6"):
            return [KEYBOARD_6]
        case .Char("7"):
            return [KEYBOARD_7]
        case .Char("8"):
            return [KEYBOARD_8]
        case .Char("9"):
            return [KEYBOARD_9]
        case .Char("a"):
            return [KEYBOARD_a]
        case .Char("b"):
            return [KEYBOARD_b]
        case .Char("c"):
            return [KEYBOARD_c]
        case .Char("d"):
            return [KEYBOARD_d]
        case .Char("e"):
            return [KEYBOARD_e]
        case .Char("f"):
            return [KEYBOARD_f]
        case .Char("g"):
            return [KEYBOARD_g]
        case .Char("h"):
            return [KEYBOARD_h]
        case .Char("i"):
            return [KEYBOARD_i]
        case .Char("j"):
            return [KEYBOARD_j]
        case .Char("k"):
            return [KEYBOARD_k]
        case .Char("l"):
            return [KEYBOARD_l]
        case .Char("m"):
            return [KEYBOARD_m]
        case .Char("n"):
            return [KEYBOARD_n]
        case .Char("o"):
            return [KEYBOARD_o]
        case .Char("p"):
            return [KEYBOARD_p]
        case .Char("q"):
            return [KEYBOARD_q]
        case .Char("r"):
            return [KEYBOARD_r]
        case .Char("s"):
            return [KEYBOARD_s]
        case .Char("t"):
            return [KEYBOARD_t]
        case .Char("u"):
            return [KEYBOARD_u]
        case .Char("v"):
            return [KEYBOARD_v]
        case .Char("w"):
            return [KEYBOARD_w]
        case .Char("x"):
            return [KEYBOARD_x]
        case .Char("y"):
            return [KEYBOARD_y]
        case .Char("z"):
            return [KEYBOARD_z]
        case .Return:
            return [KEYBOARD_Enter]
        case .Shift:
            return [KEYBOARD_Caps]
        case .SymbolShift:
            return [KEYBOARD_Symbol]

        case .Char("."):
            return [KEYBOARD_Symbol, KEYBOARD_m]
        case .Char(";"):
            return [KEYBOARD_Symbol, KEYBOARD_o]
        case .Char(","):
            return [KEYBOARD_Symbol, KEYBOARD_n]
        case .Char("\""):
            return [KEYBOARD_Symbol, KEYBOARD_p]
        case .Break:
            return [KEYBOARD_Caps, KEYBOARD_space]
        case .CursorLeft:
            return [KEYBOARD_Caps, KEYBOARD_5]
        case .CursorDown:
            return [KEYBOARD_Caps, KEYBOARD_6]
        case .CursorUp:
            return [KEYBOARD_Caps, KEYBOARD_7]
        case .CursorRight:
            return [KEYBOARD_Caps, KEYBOARD_8]
        case .Delete:
            return [KEYBOARD_Caps, KEYBOARD_0]
        case .Edit:
            return [KEYBOARD_Caps, KEYBOARD_1]
        case .ExtendedMode:
            return [KEYBOARD_Caps, KEYBOARD_Symbol]
        case .Graphics:
            return [KEYBOARD_Caps, KEYBOARD_9]
        case .InverseVideo:
            return [KEYBOARD_Caps, KEYBOARD_4]
        case .ShiftLock:
            return [KEYBOARD_Caps, KEYBOARD_2]
        case .TrueVideo:
            return [KEYBOARD_Caps, KEYBOARD_3]

        default:
            return nil
        }
    }
    
    override public func handle(event: Event) -> Bool {
        switch event {
        case .key(let key, let pressed):
            if let keyNames = fuseKeys(for: key) {
                if pressed {
                    pressedKeys.add(key)
                    if pressedKeys.count(for: key) == 1 {
                        for keyName in keyNames {
                            keyboard_press(keyName)
                        }
                    }
                }
                else {
                    pressedKeys.remove(key)
                    if pressedKeys.count(for: key) == 0 {
                        for keyName in keyNames {
                            keyboard_release(keyName)
                        }
                    }
                }
            }
            
        case .quit:
            fuse_exiting = 1;
            
        default:
            break
        }
        
        return true
    }
    
    override public func start() {
        guard let modelName = fuseName(for: machine.specification.computer.viceMachineModel) else { return } // TODO: close view
        var args = [
            "fuse",
            "--no-interface2", "--no-zxprinter",
//            "--no-traps", "--no-fastload", "--no-accelerate-loader",
            "--machine", modelName
        ]
        if let tapeImage = machine.tapeImages.first, let url = tapeImage.url {
            args.append("--tape")
            args.append(url.path)
        }
        fuseThread?.args = args
        fuseThread?.start()
    }
    
    public override func set(borderMode: MachineConfigOld.BorderMode) {
        fuseThread?.newBorderMode = borderMode.cValue
    }
}

